pkg/apis/cartographer/v1alpha1/workload_helpers.go

```diff
@@ -200,6 +202,24 @@ func (w *WorkloadSpec) MergeParams(key string, value interface{}) {
 	w.Params = append(w.Params, param)
 }
 
+func (w *WorkloadSpec) MergeComplexParams(key string, value interface{}) {
+	var obj interface{}
+	switch value.(type) {
+	case string:
+		s := value.(string)
+		s = strings.ReplaceAll(s, `\n`, "\n")
+		s = strings.ReplaceAll(s, `\t`, "\t")
```

Comment:
several local tests showed `\n` and `\t` being passed as two different characters instead of one, those values are directly from *cobra* API

Reply:
Building the code of the *branch*, for instance, and creating/updating workloads without those `replace`:
```bash
$ tanzu apps wld create tanzu-java-web-app \
--image gcr.io/dalfonso-tanzu-dev-frmwrk/tanzu-java-web-app-2-6-default:latest \
--param-yaml ports_yaml="name: smtp\nport: 1026"

error converting YAML to JSON: yaml: mapping values are not allowed in this context
```

```bash
$ tanzu apps wld create tanzu-java-web-app \ 
--image gcr.io/dalfonso-tanzu-dev-frmwrk/tanzu-java-web-app-2-6-default:latest \
--param-yaml ports_json="{\n\"name\": \"smtp\",\n \"port\": 1026\n}"

Create workload:
      1 + |---
      2 + |apiVersion: carto.run/v1alpha1
      3 + |kind: Workload
      4 + |metadata:
      5 + |  name: tanzu-java-web-app
      6 + |  namespace: default
      7 + |spec:
      8 + |  image: gcr.io/dalfonso-tanzu-dev-frmwrk/tanzu-java-web-app-2-6-default:latest
      9 + |  params:
     10 + |  - name: ports_json
     11 + |    value:
     12 + |      \n "port": 1026\n
     13 + |      \n"name": smtp
```

Reply:
You're hacking around incorrectly escaped values, partial encoding/decoding is a great way to introduce bugs and security vulnerabilities into a system.

Different shells will have slightly different escaping semantics, but for bash:

```bash
bash-3.2$ echo "foo\nbar"
foo\nbar
bash-3.2$ echo 'foo\nbar'
foo\nbar
bash-3.2$ echo $'foo\nbar'
foo
bar
```

See https://unix.stackexchange.com/questions/20035/how-to-add-newlines-into-variables-in-bash-script.


(all executions below removed the replaces)

The shell command you're looking for is:

```
tanzu apps wld create tanzu-java-web-app \
--image gcr.io/dalfonso-tanzu-dev-frmwrk/tanzu-java-web-app-2-6-default:latest \
--param-yaml ports_json=$'{\n\"name\": \"smtp\",\n \"port\": 1026\n}'
Create workload:
      1 + |---
      2 + |apiVersion: carto.run/v1alpha1
      3 + |kind: Workload
      4 + |metadata:
      5 + |  name: tanzu-java-web-app
      6 + |  namespace: default
      7 + |spec:
      8 + |  image: gcr.io/dalfonso-tanzu-dev-frmwrk/tanzu-java-web-app-2-6-default:latest
      9 + |  params:
     10 + |  - name: ports_json
     11 + |    value:
     12 + |      name: smtp
     13 + |      port: 1026
```

That sample is a bit contrived since all of the escape character are optional and this works equally as well.

```
tanzu apps wld create tanzu-java-web-app \
--image gcr.io/dalfonso-tanzu-dev-frmwrk/tanzu-java-web-app-2-6-default:latest \
--param-yaml ports_json='{"name": "smtp", "port": 1026}' 
Create workload:
      1 + |---
      2 + |apiVersion: carto.run/v1alpha1
      3 + |kind: Workload
      4 + |metadata:
      5 + |  name: tanzu-java-web-app
      6 + |  namespace: default
      7 + |spec:
      8 + |  image: gcr.io/dalfonso-tanzu-dev-frmwrk/tanzu-java-web-app-2-6-default:latest
      9 + |  params:
     10 + |  - name: ports_json
     11 + |    value:
     12 + |      name: smtp
     13 + |      port: 1026
```

Here's what it should look like in yaml:

```
tanzu apps wld create tanzu-java-web-app \
--image gcr.io/dalfonso-tanzu-dev-frmwrk/tanzu-java-web-app-2-6-default:latest \
--param-yaml ports_yaml=$'name: smtp\nport: 1026'   
Create workload:
      1 + |---
      2 + |apiVersion: carto.run/v1alpha1
      3 + |kind: Workload
      4 + |metadata:
      5 + |  name: tanzu-java-web-app
      6 + |  namespace: default
      7 + |spec:
      8 + |  image: gcr.io/dalfonso-tanzu-dev-frmwrk/tanzu-java-web-app-2-6-default:latest
      9 + |  params:
     10 + |  - name: ports_yaml
     11 + |    value:
     12 + |      name: smtp
     13 + |      port: 1026
```

Reply:
Thank you for the explanation, for avoiding it, the replacements were removed